--- a/arch/arm/boot/dts/overlays/Makefile	2016-12-19 04:12:18.044752229 +0000
+++ b/arch/arm/boot/dts/overlays/Makefile	2016-12-19 04:17:23.016020023 +0000
@@ -77,6 +77,7 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	sdio.dtbo \
 	sdio-1bit.dtbo \
 	sdtweak.dtbo \
+	slice.dtbo \
 	smi.dtbo \
 	smi-dev.dtbo \
 	smi-nand.dtbo \
@@ -96,7 +97,8 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	vga666.dtbo \
 	w1-gpio.dtbo \
 	w1-gpio-pullup.dtbo \
-	wittypi.dtbo
+	wittypi.dtbo \
+	ws2812.dtbo
 
 targets += dtbs dtbs_install
 targets += $(dtbo-y)
