--- a/arch/arm/boot/dts/overlays/README	2016-10-01 08:57:18.587961262 +0100
+++ b/arch/arm/boot/dts/overlays/README	2016-10-01 09:25:33.254153689 +0100
@@ -1220,6 +1220,11 @@ Params: speed                   Display
             Old touch display:
                 dtoverlay=tinylcd35,touch,touchgpio=3
 
+Name:   ws2812
+Info:   WS2812 LED driver using Raspberry Pi PWM interface
+Load:   dtoverlay=ws2812,<param>=<val>
+Params: invert                   Enable inversion of the PWM output
+        num_leds                 Configure number of LEDs
 
 Name:   uart1
 Info:   Enable uart1 in place of uart0
