From 671e82bc3ab4ca55a3fd52e36ce2e0aaecfcdd53 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Thu, 14 Feb 2019 06:42:34 +0000
Subject: [PATCH 07/19] Implement free_slab

---
 pan_nondrm.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index f48e105..04c2059 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -296,6 +296,18 @@ panfrost_nondrm_allocate_slab(struct panfrost_screen *screen,
         mem->stack_bottom = 0;
 }
 
+
+static void
+panfrost_nondrm_free_slab(struct panfrost_screen *screen, struct panfrost_memory *mem)
+{
+        /* For SAME_VA, we simply munmap */
+
+        if (munmap((void *) (uintptr_t) mem->gpu, mem->size)) {
+                perror("munmap");
+                abort();
+        }
+}
+
 static void
 panfrost_nondrm_enable_counters(struct panfrost_screen *screen)
 {
@@ -315,6 +327,7 @@ panfrost_nondrm_enable_counters(struct panfrost_screen *screen)
         }
 }
 
+
 struct panfrost_driver *
 panfrost_create_nondrm_driver(int fd)
 {
@@ -329,6 +342,7 @@ panfrost_create_nondrm_driver(int fd)
 	driver->base.submit_vs_fs_job = panfrost_nondrm_submit_vs_fs_job;
 	driver->base.force_flush_fragment = panfrost_nondrm_force_flush_fragment;
 	driver->base.allocate_slab = panfrost_nondrm_allocate_slab;
+	driver->base.free_slab = panfrost_nondrm_free_slab;
 	driver->base.enable_counters = panfrost_nondrm_enable_counters;
 
         ret = ioctl(fd, KBASE_IOCTL_VERSION_CHECK, &version);
-- 
2.17.1

