From 9f3b80d4c470f2272af1e7a6c30534e5d200fa17 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Tue, 12 Mar 2019 04:16:17 +0000
Subject: [PATCH 18/19] Use size

---
 pan_nondrm.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 4eeebb7..273d70b 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -129,7 +129,7 @@ panfrost_nondrm_import_bo(struct panfrost_screen *screen, struct winsys_handle *
         assert(ret == 0);
 
         bo->imported = true;
-        bo->imported_size = framebuffer_import.out.va_pages * 4096;
+        bo->size[0] = framebuffer_import.out.va_pages * 4096;
         
         return bo;
 }
@@ -355,12 +355,12 @@ panfrost_nondrm_free_imported_bo(struct panfrost_screen *screen, struct panfrost
         
         /* ..we also munmap both the Mali and DRM side bits */
 
-        if (munmap((void *) (uintptr_t) bo->gpu[0], bo->imported_size)) {
+        if (munmap((void *) (uintptr_t) bo->gpu[0], bo->size[0])) {
                 perror("munmap");
                 abort();
         }
 
-        if (munmap((void *) (uintptr_t) bo->cpu[0], bo->imported_size)) {
+        if (munmap((void *) (uintptr_t) bo->cpu[0], bo->size[0])) {
                 perror("munmap");
                 abort();
         }
-- 
2.17.1

