From 3a0c7149ee9aa5128f56a47ba0a978f3e79dd9c0 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Sun, 10 Feb 2019 20:10:53 +0000
Subject: [PATCH 06/12] t6xx: Cast uintptr_t for 32-bit

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
---
 pan_nondrm.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 2f95d82..f48e105 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -169,7 +169,7 @@ panfrost_nondrm_submit_vs_fs_job(struct panfrost_context *ctx, bool has_draws, b
                 {
                         .jc = panfrost_fragment_job(ctx),
                         .nr_extres = 1,
-                        .extres_list = (u64)framebuffer,
+                        .extres_list = (u64) (uintptr_t) framebuffer,
                         .atom_number = allocate_atom(),
                         .core_req = BASE_JD_REQ_FS,
                 },
@@ -192,7 +192,7 @@ panfrost_nondrm_submit_vs_fs_job(struct panfrost_context *ctx, bool has_draws, b
         for (int i = 0; i < 2; ++i)
                 atoms[i].compat_core_req = atoms[i].core_req;
 
-        mali_ptr addr = (mali_ptr) (atoms + (has_draws ? 0 : 1));
+        mali_ptr addr = (mali_ptr) (uintptr_t) (atoms + (has_draws ? 0 : 1));
         unsigned nr_atoms = has_draws ? 2 : 1;
 
         struct kbase_ioctl_job_submit submit = {
@@ -290,7 +290,7 @@ panfrost_nondrm_allocate_slab(struct panfrost_screen *screen,
                         perror("mmap");
                         abort();
                 }
-                mem->gpu = (mali_ptr)mem->cpu;
+                mem->gpu = (mali_ptr) (uintptr_t) mem->cpu;
         }
 
         mem->stack_bottom = 0;
-- 
2.17.1

