From 819cdcc334545b3d8438e89e7228c50f6a9cf815 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Wed, 27 Feb 2019 00:30:06 +0000
Subject: [PATCH 12/12] Don't needlessly set extres

---
 pan_nondrm.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 44b7a82..0c6efea 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -172,13 +172,16 @@ panfrost_nondrm_submit_vs_fs_job(struct panfrost_context *ctx, bool has_draws, b
                 },
                 {
                         .jc = panfrost_fragment_job(ctx),
-                        .nr_extres = 1,
-                        .extres_list = (u64) (uintptr_t) framebuffer,
                         .atom_number = allocate_atom(),
                         .core_req = BASE_JD_REQ_FS,
                 },
         };
 
+        if (is_scanout) {
+                atoms[1].nr_extres = 1;
+                atoms[1].extres_list = (u64) (uintptr_t) framebuffer;
+        }
+
         if (screen->last_fragment_id != -1) {
                 atoms[0].pre_dep[0].atom_id = screen->last_fragment_id;
                 atoms[0].pre_dep[0].dependency_type = BASE_JD_DEP_TYPE_ORDER;
-- 
2.17.1

