From 59f0a2d48c85370a34a9c8ca6be49061344909e4 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Fri, 22 Feb 2019 23:43:50 +0000
Subject: [PATCH 10/19] Hardcode page

---
 include/mali_base_kernel.h | 25 ++++++++-----------------
 1 file changed, 8 insertions(+), 17 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/include/mali_base_kernel.h b/src/gallium/drivers/panfrost/nondrm/include/mali_base_kernel.h
index 297df8b..5e413c2 100644
--- a/src/gallium/drivers/panfrost/nondrm/include/mali_base_kernel.h
+++ b/src/gallium/drivers/panfrost/nondrm/include/mali_base_kernel.h
@@ -54,27 +54,18 @@ typedef struct base_mem_handle {
 
 #define BASE_MAX_COHERENT_GROUPS 16
 
-#if defined CDBG_ASSERT
-#define LOCAL_ASSERT CDBG_ASSERT
-#elif defined KBASE_DEBUG_ASSERT
-#define LOCAL_ASSERT KBASE_DEBUG_ASSERT
-#else
-#error assert macro not defined!
+#define LOCAL_ASSERT assert
+
+/* Assume 4096-byte pages */
+#ifndef PAGE_MASK
+#define PAGE_MASK (~4095)
+#endif
+#ifndef PAGE_SHIFT
+#define PAGE_SHIFT (12)
 #endif
 
-#if defined(PAGE_MASK) && defined(PAGE_SHIFT)
 #define LOCAL_PAGE_SHIFT PAGE_SHIFT
 #define LOCAL_PAGE_LSB ~PAGE_MASK
-#else
-#include <osu/mali_osu.h>
-
-#if defined OSU_CONFIG_CPU_PAGE_SIZE_LOG2
-#define LOCAL_PAGE_SHIFT OSU_CONFIG_CPU_PAGE_SIZE_LOG2
-#define LOCAL_PAGE_LSB ((1ul << OSU_CONFIG_CPU_PAGE_SIZE_LOG2) - 1)
-#else
-#error Failed to find page size
-#endif
-#endif
 
 /**
  * @addtogroup base_user_api User-side Base APIs
-- 
2.17.1

