From b6b210a3ec74de04288a425624c357d4599d2d04 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Thu, 7 Feb 2019 05:49:10 +0000
Subject: [PATCH 03/19] Remove pandev_standard_allocate

---
 pan_nondrm.c | 18 +++---------------
 1 file changed, 3 insertions(+), 15 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 8005647..5c087fa 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -85,14 +85,6 @@ pandev_general_allocate(int fd, int va_pages, int commit_pages,
         return 0;
 }
 
-static int
-pandev_standard_allocate(int fd, int va_pages, int flags,
-                         u64 *out, int *out_flags)
-{
-        return pandev_general_allocate(fd, va_pages, va_pages, 0, flags, out,
-                                       out_flags);
-}
-
 static struct panfrost_bo *
 panfrost_nondrm_create_bo(struct panfrost_screen *screen, const struct pipe_resource *template)
 {
@@ -454,13 +446,9 @@ panfrost_nondrm_allocate_slab(struct panfrost_screen *screen,
         if (same_va)
                 flags |= BASE_MEM_SAME_VA;
 
-        if (commit_count || extent)
-                pandev_general_allocate(nondrm->fd, pages,
-                                        commit_count,
-                                        extent, flags, &mem->gpu, &out_flags);
-        else
-                pandev_standard_allocate(nondrm->fd, pages, flags, &mem->gpu,
-                                         &out_flags);
+	pandev_general_allocate(nondrm->fd, pages,
+				commit_count ? commit_count : pages,
+				extent, flags, &mem->gpu, &out_flags);
 
         mem->size = pages * 4096;
 
-- 
2.17.1

