From d0166876d097833fe73e360eb5cb45f6cfb30bf7 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Tue, 12 Mar 2019 04:16:47 +0000
Subject: [PATCH 19/19] For compat with 32-bit, always set SAME_VA

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
---
 pan_nondrm.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 273d70b..f347c4e 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -295,8 +295,7 @@ panfrost_nondrm_allocate_slab(struct panfrost_screen *screen,
         if (flags & BASE_MEM_PROT_GPU_EX)
                 flags &= ~BASE_MEM_PROT_GPU_WR;
 
-        if (same_va)
-                flags |= BASE_MEM_SAME_VA;
+        flags |= BASE_MEM_SAME_VA;
 
 	pandev_general_allocate(nondrm->fd, pages,
 				commit_count ? commit_count : pages,
-- 
2.17.1

