From e46704b28d7c928680de03419f63ce6917bc902d Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Sun, 10 Mar 2019 19:15:33 +0000
Subject: [PATCH 17/19] Fix pointer size casting warning

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
---
 pan_nondrm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
index 7873183..4eeebb7 100644
--- a/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
+++ b/src/gallium/drivers/panfrost/nondrm/pan_nondrm.c
@@ -405,7 +405,7 @@ panfrost_nondrm_query_gpu_version(struct panfrost_screen *screen)
 	char *ptr;
 	int size;
 
-        get_gpuprops.buffer = buf;
+        get_gpuprops.buffer = (u64) (uintptr_t) buf;
         get_gpuprops.size = sizeof(buf);
         size = pandev_ioctl(nondrm->fd, KBASE_IOCTL_GET_GPUPROPS, &get_gpuprops);
         if (size < 1 || size > get_gpuprops.size) {
-- 
2.17.1

