From 3cd380e2a2f209294e9777c85c0ea5cea0533310 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Tue, 6 Nov 2018 18:10:41 -0800
Subject: [PATCH 078/151] Fix error on some compilers

---
 src/gallium/drivers/panfrost/pan_screen.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/gallium/drivers/panfrost/pan_screen.c b/src/gallium/drivers/panfrost/pan_screen.c
index 6dced8c..62bee33 100644
--- a/src/gallium/drivers/panfrost/pan_screen.c
+++ b/src/gallium/drivers/panfrost/pan_screen.c
@@ -580,6 +580,7 @@ panfrost_resource_from_handle(struct pipe_screen *screen,
                               unsigned usage)
 {
 	assert(0);
+        return NULL;
 }
 
 
-- 
2.7.4

