From ae24026e1676ce61789e9bb6dfff00e4a0872999 Mon Sep 17 00:00:00 2001
From: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Date: Wed, 7 Nov 2018 08:41:28 -0800
Subject: [PATCH 085/151] Disable viewport insanity

---
 src/gallium/drivers/panfrost/pan_context.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/gallium/drivers/panfrost/pan_context.c b/src/gallium/drivers/panfrost/pan_context.c
index d45d029..9004233 100644
--- a/src/gallium/drivers/panfrost/pan_context.c
+++ b/src/gallium/drivers/panfrost/pan_context.c
@@ -2574,12 +2574,14 @@ panfrost_set_viewport_states(struct pipe_context *pipe,
 
         ctx->pipe_viewport = *viewports;
 
+#if 0
         /* TODO: What if not centered? */
         float w = abs(viewports->scale[0]) * 2.0;
         float h = abs(viewports->scale[1]) * 2.0;
 
         ctx->viewport.viewport1[0] = MALI_POSITIVE((int) w);
         ctx->viewport.viewport1[1] = MALI_POSITIVE((int) h);
+#endif
 }
 
 static void
-- 
2.7.4

