From ab92302d62ec75dfb06ff38f4294e3999a84d5fa Mon Sep 17 00:00:00 2001
From: Tomeu Vizoso <tomeu.vizoso@collabora.com>
Date: Tue, 4 Dec 2018 08:22:56 +0100
Subject: [PATCH 132/151] panfrost: Add uapi include

Signed-off-by: Tomeu Vizoso <tomeu.vizoso@collabora.com>
---
 src/gallium/drivers/panfrost/meson.build | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/gallium/drivers/panfrost/meson.build b/src/gallium/drivers/panfrost/meson.build
index 1950b1e..4742429 100644
--- a/src/gallium/drivers/panfrost/meson.build
+++ b/src/gallium/drivers/panfrost/meson.build
@@ -43,6 +43,7 @@ inc_panfrost = [
   inc_common,
   inc_gallium,
   inc_gallium_aux,
+  inc_drm_uapi,
   inc_include,
   inc_src,
   include_directories('midgard'),
-- 
2.7.4

