--- a/sound/soc/bcm/Kconfig	2016-10-01 08:28:02.404275313 +0100
+++ b/sound/soc/bcm/Kconfig	2016-10-01 08:28:48.197310044 +0100
@@ -121,3 +121,10 @@ config SND_BCM2708_SOC_ALLO_PIANO_DAC
 	select SND_SOC_PCM512x_I2C
 	help
 	  Say Y or M if you want to add support for Allo Piano DAC.
+
+config SND_BCM2708_SOC_SLICE
+	tristate "Support for Slice on-board sound"
+	depends on SND_BCM2708_SOC_I2S || SND_BCM2835_SOC_I2S
+	select SND_SOC_CS4265
+	help
+	  Say Y or M if you want to add support for Slice on-board sound.
